File: GahPlugin/src/android/GahPlugin.java
package com.gemalto.plugin.gah;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.json.JSONArray;
import org.json.JSONException;

/**
 * GAH plugin wrapping the GAH SDK.
 */
public class GahPlugin extends CordovaPlugin {
    private GAHRiskEngineIntegration mEngineIntegration;

    /**
     * {@inheritDoc}
     */
    @Override
    protected void pluginInitialize() {
        super.pluginInitialize();

        mEngineIntegration = new GAHRiskEngineIntegration(cordova.getActivity().getApplicationContext());
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean execute(final String action, final JSONArray data, final CallbackContext callbackContext)
            throws JSONException {
        if (action.equals("initialize")) {
            final String gahUrl = data.getString(0);
            final String tmxOrgId = data.getString(1);
            final String tmxFpServerUrl = data.getString(2);

            if (gahUrl == null || gahUrl.isEmpty()) {
                callbackContext.error("Missing mandatory parameter: GAH URL");
                return false;
            }

            mEngineIntegration.initializeGAH(gahUrl, tmxOrgId, tmxFpServerUrl);
            callbackContext.success("initialization ok");
        } else if (action.equals("visitId")) {
            mEngineIntegration.requestVisitId(callbackContext);
        } else if (action.equals("startPrefetch")) {
            mEngineIntegration.startPrefetchCollection();
        } else if (action.equals("stopPrefetch")) {
            mEngineIntegration.stopPrefetchCollection();
            return true;
        } else if (action.equals("clearTransaction")) {
            mEngineIntegration.clearTransactionData();
            return true;
        } else if (action.equals("isSignalCollectionCompleted")) {
            mEngineIntegration.isSignalCollectionCompleted(callbackContext);
            return true;
        } else if (action.equals("setTransactionAsCritical")) {
            mEngineIntegration.setTransactionAsCritical();
        } else {
            callbackContext.error("undefined action");
            return false;
        }

        return true;
    }
}

File: GahPlugin/src/ios/GahPlugin.m
#import "GahPlugin.h"
#import "GAHRiskEngineIntegration.h"

@interface GahPlugin()

@property(nonatomic, strong) GAHRiskEngineIntegration* integrationEngine;

@end

@implementation GahPlugin

#pragma mark override

- (void)pluginInitialize {
    self.integrationEngine = [[GAHRiskEngineIntegration alloc] init];
}

#pragma mark public

- (void)initialize:(CDVInvokedUrlCommand*)command {
    NSString* gahUrl = (NSString*)[command.arguments objectAtIndex:0];
    NSString* tmxOrgId = (NSString*)[command.arguments objectAtIndex:1];
    NSString* tmxFpServerUrl = (NSString*)[command.arguments objectAtIndex:2];
    
    if (gahUrl == nil || gahUrl.length == 0) {
        CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Missing mandatory parameter: GAH URL"];
        [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
    }
    
    [_integrationEngine initializeGAH:gahUrl tmxOrgId:tmxOrgId tmxFpServerUrl:tmxFpServerUrl];
    
    CDVPluginResult* result = [CDVPluginResult resultWithStatus:CDVCommandStatus_OK messageAsString:@"Initialization ok"];
    [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
}

- (void)visitId:(CDVInvokedUrlCommand*)command {
    [_integrationEngine requestVisitID:self.commandDelegate callbackId:command.callbackId];
}

- (void)startPrefetch:(CDVInvokedUrlCommand*)command {
    [_integrationEngine startPrefetchingCollection];
}

- (void)stopPrefetch:(CDVInvokedUrlCommand*)command {
    [_integrationEngine stopPrefetchingCollection];
}

- (void)clearTransaction:(CDVInvokedUrlCommand*)command {
    [_integrationEngine clearResources];
}

- (void)isSignalCollectionCompleted:(CDVInvokedUrlCommand*)command {
    [_integrationEngine isSignalCollectionCompleted:self.commandDelegate callbackId:command.callbackId];
}

- (void)setTransactionAsCritical:(CDVInvokedUrlCommand*)command {
    [_integrationEngine setTransactionAsCritical];
}

@end
